### PR TITLE
Don't instantiate so many copies of drop_in_place

### PR DESCRIPTION
**Pull request by matthewjasper**
Split out from #66703. 

r? @ghost

### CODE REVIEW

File: src/librustc/ty/instance.rs
Comment (bjorn3) [inline src/librustc/ty/instance.rs]:
```suggestion
    /// The `DefId` is for `real_drop_in_place`. The `Option<Ty>` is `None` for empty drop glue.
```

**Comment by matthewjasper** [comment]
I've removed the normalize change - it appears to be a perf regression
@bors try @rust-timer queue

**Comment by matthewjasper** [comment]
...and I've brought it back when `Reveal::All` is set...
@bors try @rust-timer queue

**Comment by matthewjasper** [comment]
I guess the first version of this was fine...

This is fairly hacky, but it gives some significant wins, albeit in a case (release incremental) that maybe isn't considered as important as others. (https://github.com/rust-lang/rust/pull/67332#issuecomment-565900248)

cc @rust-lang/wg-compiler-performance - does anyone want to review this.

**Comment by michaelwoerister** [comment]
Thanks for the PR, @matthewjasper! The wins are significant indeed. Let me take a look at how hacky the implementation really is. Since this only affects incremental-opt builds, I don't think there is a need to rush this.

**Comment by matthewjasper** [comment]
There's also the CI failure which I'll need to address first.

File: src/librustc/ty/util.rs
Comment (eddyb) [inline src/librustc/ty/util.rs]: IIRC the approach `ty::layout` takes is to check `normalized == ty`, but this is just nit-picking.

**Comment by JohnTitor** [comment]
Ping from triage: @matthewjasper any updates on this? Also, there are some merge conflicts.

**Comment by matthewjasper** [comment]
blocked on #68679

File: src/librustc/ty/instance.rs
Comment (RalfJung) [inline src/librustc/ty/instance.rs]: `real_drop_in_place` is gone (https://github.com/rust-lang/rust/pull/68335), it's just `drop_in_place` now.

File: src/librustc_mir/monomorphize/partitioning.rs
Comment (RalfJung) [inline src/librustc_mir/monomorphize/partitioning.rs]: Some more rebase leftovers of `real_drop_in_place`.

**Comment by matthewjasper** [comment]
@eddyb @michaelwoerister this is now ready for review

**Review by eddyb — APPROVED** [review]
LGTM, r? @michaelwoerister

**Comment by michaelwoerister** [comment]
Thanks again for the PR, @matthewjasper! 

Since we have incremental ThinLTO now, hopefully that will make up for most of the potential performance loss. If we see complaints about decreased runtime performance, we could change this to just apply at opt-levels below 2. 

If I read this correctly, this puts all of drop-glue into a single object file (the "fallback" module). That looks potentially problematic because it causes that object file to be recompiled whenever a type implementing `Drop` is changed (or something like that). I'll think about this some more before r+ing.

**Comment by michaelwoerister** [comment]
OK, I'm up for merging this, the performance wins look great!

There's one optional modification that you could make, which is putting drop-glue into the module of the corresponding type. That should be possible by modifying this case:

https://github.com/rust-lang/rust/blob/d49ec090453d901ed94c60d5fb02cd5fe6c3254d/src/librustc_mir/monomorphize/partitioning.rs#L672

i.e. create the `Instance` via `Instance::resolve_drop_in_place`, extract the type param and use `characteristic_def_id_of_type()`. 

I'll leave it up to you whether you want to try this or just r=me the current state.

**Comment by bjorn3** [comment]
The single field of `InstanceDef::DropGlue` already is the type for which the instance is the drop glue.

**Comment by michaelwoerister** [comment]
Oh that's right. I forgot about that. Good catch!

**Comment by matthewjasper** [comment]
I tried putting drop glue into the type's module and found it to have worse (compiler) performance in the tests that I did.
@bors r=michaelwoerister

**Comment by michaelwoerister** [comment]
OK, sounds good then :+1:

**Pull request merged**
